[SCEV] Infer loop max trip count from memory accesses

shiltian opened the pull request:
Data references in a loop is assumed to not access elements over the statically
allocated size. We can therefore infer a loop max trip count from this undefined
behavior.

This patch is refined from the orignal one (https://reviews.llvm.org/D155049)
authored by @Peakulorain.

jdoerfert commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -249,6 +249,10 @@ static cl::opt<bool> UseContextForNoWrapFlagInference(
 
+static cl::opt<bool> UseMemoryAccessUBForBEInference(
+    "scalar-evolution-infer-max-trip-count-from-memory-access", cl::Hidden,
+    cl::desc("Infer loop max trip count from memory access"), cl::init(false));
```
Can you check the compile time impact (via @nikic tracker) when you enable this?

jdoerfert commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8191,6 +8204,153 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+  return SE->getConstant(RTy, Size);
+}
+
+static const SCEV *countItersSelfWrap(const SCEV *V, ScalarEvolution *SE) {
```
Documentation.

jdoerfert commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8191,6 +8204,153 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+  return SE->getCouldNotCompute();
+}
+
+/// Return the smaller one of the wraps that will occur in the indexes.
```
/// Return the smallest number of iterations until any index in the GEP \p Ptr wraps, or ...

jdoerfert commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8191,6 +8204,153 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+    if (isa<ZExtInst>(V) || isa<SExtInst>(V))
+      V = cast<Instruction>(Index)->getOperand(0);
+    const SCEV *Count = countItersSelfWrap(SE->getSCEV(V), SE);
+    if (!isa<SCEVCouldNotCompute>(Count))
```
How can you just ignore things here, that seems wrong. Or you use CouldNotCompute to indicate that there is no wrap, which is not advisable.

jdoerfert commented on llvm/test/Analysis/ScalarEvolution/infer-trip-count-idx-wrap.ll:
```diff
@@ -0,0 +1,110 @@
+; CHECK-NEXT:  Loop %for.body: Predicated backedge-taken count is (-1 + %len)
+; CHECK-NEXT:   Predicates:
+; CHECK:       Loop %for.body: Trip multiple is 1
+; CHECK-NEXT:  Loop %for.body: Small constant max trip is 21
```
Shouldn't it be 20? Where is the off by one coming from?

jdoerfert commented on llvm/test/Analysis/ScalarEvolution/infer-trip-count.ll:
```diff
@@ -0,0 +1,191 @@
+  ret void
+}
+
+define void @ComputeMaxTripCountFromMultiDemArray(i32 signext %len) {
```
-Dem +Dim

nikic reviewed (commented):
This still has the weird "self wrap" code instead of checking addrec nowrap flags.

fmayer commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8191,6 +8204,132 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+    return;
+
+  const Function *F = LoopLatch->getParent();
+  if (F->hasFnAttribute(Attribute::SanitizeAddress) ||
```
Could you add a comment why we are doing this?

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8191,6 +8204,133 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+  for (Instruction *I : MemInsts) {
+    Value *Ptr = getLoadStorePointerOperand(I);
+    assert(Ptr && "empty pointer operand");
+    if (!checkIndexWrap(Ptr, this))
+      continue;
+    auto *AddRec = dyn_cast<SCEVAddRecExpr>(getSCEV(Ptr));
+    if (!AddRec || !AddRec->isAffine())
```
```suggestion
    auto *AddRec = dyn_cast<SCEVAddRecExpr>(getSCEV(Ptr));
    if (!AddRec || !AddRec->isAffine() || !AddRec->hasNoUnsignedWrap())
```
Does this work? I still don't understand what exactly you're trying to achieve with checkIndexWrap().

shiltian replied:
Actually I need to refine this part. Both what we have now (`checkIndexWrap`) and what you suggested don't work for the following case:
```
define void @ComputeMaxTripCountFromArrayIdxWrap(i32 signext %len) {
entry:
  %a = alloca [257 x i32], align 4
  %cmp4 = icmp sgt i32 %len, 0
  br i1 %cmp4, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:
  br label %for.body

for.cond.cleanup.loopexit:
  br label %for.cond.cleanup

for.cond.cleanup:
  ret void

for.body:
  %iv = phi i8 [ %inc, %for.body ], [ 0, %for.body.preheader ]
  %idxprom = zext i8 %iv to i64
  %arrayidx = getelementptr inbounds [257 x i32], [257 x i32]* %a, i64 0, i64 %idxprom
  store i32 0, i32* %arrayidx, align 4
  %inc = add nuw nsw i8 %iv, 1
  %inc_zext = zext i8 %inc to i32
  %cmp = icmp slt i32 %inc_zext, %len
  br i1 %cmp, label %for.body, label %for.cond.cleanup.loopexit
}
```
Here `%iv` is `i8` that is in the range of 0-255, so the max trip count in this case can not be deduced. All those wrap flags fail to tell `%iv` can actually wrap. We will probably need to adopt what is used in the original patch:
```
static const SCEV *howManyItersSelfWrap(const SCEV *V, ScalarEvolution *SE) {
  if (auto *AddRec = dyn_cast<SCEVAddRecExpr>(V)) {
    const SCEV *CUpper = SE->getConstant(SE->getUnsignedRangeMax(V));
    const SCEV *CLower = SE->getConstant(SE->getUnsignedRangeMin(V));
    const SCEV *Limit = SE->getMinusSCEV(CUpper, CLower);
    const SCEV *Step = AddRec->getStepRecurrence(*SE);
    return SE->getUDivCeilSCEV(Limit, Step);
  }
  return SE->getCouldNotCompute();
}

/// Returns the smaller one of the wraps that will occur in the indexes.
static const SCEV *getSmallCountOfIdxSelfWrap(Value *Ptr, ScalarEvolution *SE) {
  auto *PtrGEP = dyn_cast<GetElementPtrInst>(Ptr);
  if (!PtrGEP)
    return SE->getCouldNotCompute();

  SmallVector<const SCEV *> CountColl;
  for (Value *Index : PtrGEP->indices()) {
    Value *V = Index;
    if (isa<ZExtInst>(V) || isa<SExtInst>(V))
      V = cast<Instruction>(Index)->getOperand(0);
    const SCEV *Count = howManyItersSelfWrap(SE->getSCEV(V), SE);
    if (!isa<SCEVCouldNotCompute>(Count)) {
      CountColl.push_back(Count);
    }
  }

  if (CountColl.empty())
    return SE->getCouldNotCompute();

  return SE->getUMinFromMismatchedTypes(CountColl);
}

...

    ConstantInt *WrapVC = IdxWrapMap[Rec]->getValue();
    ConstantInt *InferVC = InferCount->getValue();
    if (InferVC->getValue().getZExtValue() > WrapVC->getValue().getZExtValue())
      continue;
```
Or is there anything else that can tell the potential wrap here?

nikic replied:
The example you shared cannot wrap due to presence of `add nuw`. If `%iv` wrapped, a poison value would reach the latch condition, which would result in undefined behavior. As such, we can assume that `%iv` does not wrap, and as such `%len` cannot be larger than 256. Am I missing something here?

shiltian replied:
Okay, yeah, that makes sense. However, with what we have right now, we will get the max trip count 257, which is out of the range of `%iv`. That might be the reason that the initial patch has the "weird" get small count of index thing.

shiltian commented:
Gentle ping

jdoerfert commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8190,6 +8194,183 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+    } else {
+      auto *IndexRangeC = dyn_cast<SCEVConstant>(IndexRange);
+      if (!IndexRangeC)
+        continue;
```
Again, this looks like an abort condition. The wrap stuff would otherwise allow some weird index range here as long as there is a second index in the gep that behaves nicely.

shiltian commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8253,6 +8261,210 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+}
+
+/// Get the underlying SCEVAddExpr from a cast expression if possible.
+const SCEV *peelCastExpr(const SCEVCastExpr *S, ScalarEvolution *SE) {
```
Do we have existing utility that can already do this?

shiltian commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8253,6 +8261,210 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+  return SE->getCouldNotCompute();
+}
+
+static Value *peelExt(Value *V) {
```
Do we have existing utility that can already do this?

shiltian commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8253,6 +8261,210 @@ ScalarEvolution::getSmallConstantTripMultiple(const Loop *L,
+  return V;
+}
+
+static bool isIndexInductionVariable(PHINode *InductionVar, Value *Index) {
```
Do we have existing utility that can already do this?

shiltian replied:
This might be too conservative. It can't handle uses such as `array[i * 2]`. What is the best way to do that?

shiltian commented on llvm/test/Analysis/ScalarEvolution/infer-trip-count.ll:
```diff
@@ -0,0 +1,191 @@
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
```
The tests need to be improved by adding:
1. Negative step
2. Multi-dim array with nested loop with both forwarded and reversed indices

shiltian commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -249,6 +250,13 @@ static cl::opt<bool> UseContextForNoWrapFlagInference(
+    "scalar-evolution-infer-max-trip-count-from-memory-access", cl::Hidden,
+    cl::desc("Infer loop max trip count from memory access"), cl::init(false));
+
+DEBUG_COUNTER(MemoryAccessCounter, "memory-access-ub",
```
This will be removed once it pass the LLVM test suite.

shiltian commented:
This version can pass llvm test suite w/ the option `-mllvm -scalar-evolution-infer-max-trip-count-from-memory-access=1`.

shiltian committed:
[SCEV] Infer loop max trip count from memory accesses

Data references in a loop is assumed to not access elements over the statically
allocated size. We can therefore infer a loop max trip count from this undefined
behavior.

This patch is refined from the orignal one (https://reviews.llvm.org/D155049)
authored by @Peakulorain.

fmayer reviewed (commented):
LGTM from sanitizer point of view

shiltian commented:
This PR doesn’t work as expected for a real-world use case.
